### DISCUSSION TITLE
Trimming with window event

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#12203] [Q&A]
I'm having a little issue to publish with trimming.

I'm binding to `WindowBase.Activated` that is not a RoutedEvent, and thus cannot use `RoutedEventTriggerBehavior`.

`csproj ` contains

```
<PropertyGroup>
  <AssemblyName>YangDownloader.Core</AssemblyName>
  <RootNamespace>HanumanInstitute.YangDownloader</RootNamespace>
</PropertyGroup>
<ItemGroup>
  <TrimmerRootDescriptor Include="Roots.xml" />
</ItemGroup>
```

`Roots.xml` contains
```
<linker>
    <assembly fullname="YangDownloader.Core" preserve="All"/>
    <assembly fullname="YoutubeExplode" preserve="All"/>
</linker>
```

Yet; it still trims the main assembly and crashes on start. What am I missing?

btw it seems the template has `Roots.xml` but not the `TrimmerRootDescriptor` entry; although I haven't double-checked

**Comment by maxkatz6** [comment]
If you need to root Activated event, then you need to add Avalonia.Base (or Controls, not remember right now) assembly to the file.

**Reply by mysteryx93** [reply]
Ah gotcha. Wouldn't have thought of that. Not working though. Tried with both `Avalonia.Base` and `Avalonia.Controls`.

    <assembly fullname="Avalonia.Base" Preserve="All" />

If the above would work, this should do the job, right?
```
<assembly fullname="Avalonia.Base">
    <type fullname="Avalonia.Controls.WindowBase">
        <event name="Activated" />
    </type>
</assembly>
```

Btw, did you make improvements to trimming in the final release? With the preview, one of the libraries was larger with 25.9mb (compressed) on Linux. With the final release, that same project takes 19.2mb. Removing the font made a difference, but here I'm seeing a greater difference.

**Reply by mysteryx93** [reply]
Adding this to the Main method, however, works

    [DynamicDependency(DynamicallyAccessedMemberTypes.PublicEvents, typeof(WindowBase))]

but I'd want to know how to make it work with the `Roots.xml` file

**Reply by mysteryx93** [reply]
With v11.0.0 final, output size is considerably smaller in Linux only, but very slightly larger in Windows and MacOS. Interesting.

**Reply by maxkatz6** [reply]
> but I'd want to know how to make it work with the Roots.xml file

Attributes are preferable anyway. Roots.xml file isn't even well documented IIRC.
